### PR TITLE
update adam documentation

### LINKED ISSUE #42843: Documentation mistake of Adam in v1.6.0?
## 📚 Documentation

The documentation of Adam since v1.6.0 suggests that it uses the weight decay fix proposed by paper "Decoupled Weight Decay Regularization":
https://github.com/pytorch/pytorch/blob/4b4273a04e566dcdf4cb96882f49c41127e8d3f1/torch/optim/adam.py#L10-L11

However I found this to not be the case. The actual weight decay in Adam is still the old one:
https://github.com/pytorch/pytorch/blob/4b4273a04e566dcdf4cb96882f49c41127e8d3f1/torch/optim/adam.py#L99-L100

In contrast, the new weight decay fix is implemented in AdamW, and explained by the AdamW doc:
https://github.com/pytorch/pytorch/blob/4b4273a04e566dcdf4cb96882f49c41127e8d3f1/torch/optim/adamw.py#L10

https://github.com/pytorch/pytorch/blob/4b4273a04e566dcdf4cb96882f49c41127e8d3f1/torch/optim/adamw.py#L73

It is likely that I misunderstood the current documentation of Adam, but I found it confusing and suggest that there is no difference between the Adam and AdamW implementation, while the difference still exists.

### PR DESCRIPTION
**Pull request by cuscada**
Possibly patching minor wording issue that confuses reader to the documentation of Adam L2 penalty, since there is no obvious change whatsoever.

Fixes #42843

### CODE REVIEW

**Comment by grrlic** [comment]
Replying to https://github.com/grrlic/pytorch/pull/1#issuecomment-674295003
> Hi @grrlic, I noticed that you reference my original issue [pytorch#42843](https://github.com/pytorch/pytorch/issues/42843) for this change. Thanks for the quick response on this!
> 
> I hope you don't mind a quick follow-up: Can you explain why the current Adam implementation stills "follows the proposal in `Decoupled Weight Decay Regularization`"? My original understanding is that only the implementation of AdamW is closely related to this paper, but I might be wrong on this too.

Hi @yuhaozhang thanks for addressing this Adam implementations issue. You made the point, it indeed can be quite confusing.

It seems the authors would deliberately separate `adamw.py` from the old `adam.py` by adding the Stepweight decay. But overall, it still moves in accordance with the three arXiv papers.

Let me know if I made wrong points.

**Comment by yuhaozhang** [comment]
Hi @grrlic, thanks for your explanations! While I do understand that the implementation of `adam.py` is highly related to the other two papers ("Adam: A Method for Stochastic Optimization:" and "On the Convergence of Adam and Beyond"), I still don't understand how it is influenced by the paper "Decoupled Weight Decay Regularization", to the extent that this paper needs to be highlighted in the documentation. To my understanding this paper proposed the weight decay fix which closely relates to the `adamw.py`. Can you further clarify this?

**Comment by vincentqb** [comment]
The change looks good to me, but I'm not sure whether it fixes #42823. Can you explain how it addresses the original issue?

**Comment by grrlic** [comment]
@yuhaozhang Sorry for just tuning in, if we see from the past commits [33640](https://github.com/pytorch/pytorch/commit/6a97777f7256b20d874352263868dda81c674b53#diff-1ad7ee25f9d26b47f9a3a1e38555caf3) and [34211](https://github.com/pytorch/pytorch/commit/6e2bb1c05442010aff90b413e21fce99f0393727#diff-1ad7ee25f9d26b47f9a3a1e38555caf3), the `adam.py` are adapting to the proposal by Ilya Loshchilov & Frank Hutter. Perhaps you have a different perspective on this? Thanks.

**Comment by grrlic** [comment]
@vincentqb Hey Vincent glad for you dropping in. I am pretty sure this PR is not in any way associated with issue #42823, but #42843 

If only the one with write access would soon resolve this PR if so far the change looks good.

**Comment by dvolgyes** [comment]
@vincentqb Any chance you could review that?
But I still think the documentation is wrong, and this wording is still misleading.
So here is the background:

Weight decay often implemented as:
```
Loss = Loss + alpha*||network_weights||^2
```
This is what is implemented in Adam too.
But the mentioned Decoupled Weight Decay (DWD) paper points out that this
approach with adaptive optimizers is actually wrong. 
(Very simply: if you reach 0, the optimizer can overhsoot because it has momentum.)
The Decoupled Weight Decay paper semantically proposes this:
1, take the optimization step
2, disable autograd
3, decrease the weights
4, enable back autograd

This is the AdamW algorithm.

The problem with the documentation that it implies that Adam is `fixed`, however it is not.
It still uses the original L2 penalty, there is no connection at all to the DWD paper, 
except that they named the old approach L2 penalty, and the new one as (fixed) weight decay.
It is misleading, and the reference to the DWD paper should be completely removed.

The current text is:
```
It has been proposed in Adam: A Method for Stochastic Optimization. The implementation of the L2 penalty follows changes proposed in Decoupled Weight Decay Regularization.
```
One alternative:
```
It has been proposed in Adam: A Method for Stochastic Optimization. The implementation uses L2 penalty.
```
Or 
```
It has been proposed in Adam: A Method for Stochastic Optimization. This implementation uses L2 penalty. See AdamW for the changes proposed in Decoupled Weight Decay Regularization paper.
```
Or something similar, but the committed change DO NOT improve the documentation, it is still wrong, and you should consider something I just mentioned a few lines above.

Anyhow, this documentation fix is quite old, it would be nice to fix it finally, could someone from Pytorch step up?
Maybe 1.9 could be released with corrected docs.
